### ISSUE TITLE
group_by filter behavior changed between 4.0.1 and 4.1.0

### ISSUE THREAD

**Issue by moozer** [jekyll/jekyll#8214]
## My Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Operating System |    debian buster        |
| `jekyll`         |  4.0.1 and 4.1.0  |
| `github-pages`   | unused     |

---

## Expected Behaviour

I have jekyll sites running on gitlab pages, and today the build failed.

I have a group_by filter to list 'tags' from a collection, and changing from 4.0.1 to 4.1.0 changed the return value.

I use `group_by` to count occurrences of each tag.

## Code Sample

I have a collection where I have "tags" array in the front matter.

```
{% assign alldocs = site.exercises %}
{% assign grouptag =  alldocs | map: 'tags' | join: ','  | split: ','  | group_by: tag %}
xx {{ grouptag }}
```

in version 4.0.1 output is
```
xx {"name"=>"tag A", "items"=>["tag A"], "size"=>1}{"name"=>"Tag B", "items"=>["Tag B"], "size"=>1}{"name"=>"Virtualbox", "items"=>["Virtualbox"], "size"=>1}{"name"=>"netcat", "items"=>["netcat"], "size"=>1}{"name"=>"whois", "items"=>["whois"], "size"=>1}{"name"=>"dig", "items"=>["dig"], "size"=>1}{"name"=>"DNSEnum", "items"=>["DNSEnum"], "size"=>1}{"name"=>"DNSRecon", "items"=>["DNSRecon"], "size"=>1}{"name"=>"Shodan", "items"=>["Shodan"], "size"=>1}{"name"=>"theHarvester", "items"=>["theHarvester"], "size"=>1}{"name"=>"Recon-ng", "items"=>["Recon-ng"], "size"=>1}{"name"=>"Nmap", "items"=>["Nmap"], "size"=>1}{"name"=>"Tor", "items"=>["Tor"], "size"=>1}{"name"=>"HackTheBox", "items"=>["HackTheBox"], "size"=>1}{"name"=>"gobuster", "items"=>["gobuster"], "size"=>1}{"name"=>"dirbuster", "items"=>["dirbuster"], "size"=>1}{"name"=>"dirb", "items"=>["dirb"], "size"=>1}{"name"=>"Nikto", "items"=>["Nikto"], "size"=>1}{"name"=>"ZAP", "items"=>["ZAP"], "size"=>1}{"name"=>"Basic Authtication", "items"=>["Basic Authtication"], "size"=>1}{"name"=>"Token Authtication", "items"=>["Token Authtication"], "size"=>1}{"name"=>"Brute-Force", "items"=>["Brute-Force"], "size"=>1}{"name"=>"Dictionary attack", "items"=>["Dictionary attack"], "size"=>1}{"name"=>"Burp", "items"=>["Burp"], "size"=>1}{"name"=>"Zap", "items"=>["Zap"], "size"=>1}{"name"=>"Hydra", "items"=>["Hydra"], "size"=>1}
```

in version 4.1.0 output is
```
xx {"name"=>"", "items"=>["tag A", "Tag B", "Virtualbox", "netcat", "whois", "dig", "DNSEnum", "DNSRecon", "Shodan", "theHarvester", "Recon-ng", "Nmap", "Tor", "HackTheBox", "gobuster", "dirbuster", "dirb", "Nikto", "ZAP", "Basic Authtication", "Token Authtication", "Brute-Force", "Dictionary attack", "Burp", "Zap", "Hydra"], "size"=>26}

```

**Comment by ashmaroli** [comment]
@moozer I agree with you that there is a behavior change.
But from your code below, it looks like the change was a good thing.
```
{% assign result = site.exercises | map: 'tags' | join: ','  | split: ','  | group_by: tag %}
```
Meaning, I'm passing the blame onto you.

---
Passing the blame isn't going to resolve the situation. So let us walk through the pipeline step-by-step:
1. `site.exercises | map: 'tags'` returns an array of *tags* registered by the collection's documents. e.g. `["tag A", "Tag B", "Virtualbox"]`
2. `["tag A", "Tag B", "Virtualbox"] | join: ','  | split: ','` essentially gives a similar result as in Step 1 because there are no tags containing literal commas
3. Finally, `["tag A", "Tag B", "Virtualbox"] | group_by: tag` returns the breaking result because:
    - `group_by` is used to group objects by their `property`.
    - `tag` is a not a valid property. `'tag'` or `"tag"` would have been a valid property.
    - every item of the input array is a string.
    - strings do not have a property named `tag` / `'tag'` / `"tag"`. So it returns an empty string.

---
*You can split your code across the `|` character into multiple expressions, assign the result to a variable and inspect them using the `inspect` filter to confirm*:
```
{% assign part1 = site.exercises | map: 'tags' %}
{{ part1 | inspect }}
{% assign part2 = site.exercises | map: 'tags' | join: ',' %}
{{ part2 | inspect }}
{% assign part3 = site.exercises | map: 'tags' | join: ','  | split: ',' %}
{{ part3 | inspect }}
```

**Comment by moozer** [comment]
Thanks for showing an interest.

> Meaning, I'm passing the blame onto you.
I won't argue too much about that, since I could not find anything in the docs about my use of group_by 

> 
> Passing the blame isn't going to resolve the situation. So let us walk through the pipeline step-by-step:
> 
> 1. `site.exercises | map: 'tags'` returns an array of _tags_ registered by the collection's documents. e.g. `["tag A", "Tag B", "Virtualbox"]`
> 2. `["tag A", "Tag B", "Virtualbox"] | join: ','  | split: ','` essentially gives a similar result as in Step 1 because there are no tags containing literal commas

True, except that I allow both
```
tags: A, B, C
```
and
```
tags:
- A
- B
- C
```

> 3. Finally, `["tag A", "Tag B", "Virtualbox"] | group_by: tag` returns the breaking result because:
>    
>    * `group_by` is used to group objects by their `property`.
>    * `tag` is a not a valid property. `'tag'` or `"tag"` would have been a valid property.
>    * every item of the input array is a string.
>    * strings do not have a property named `tag` / `'tag'` / `"tag"`. So it returns an empty string.
> 

Ok, so it is not a bug - or rather my previous usage of group_by was allowed due to a bug.

How do I count the number of unique tags in the list?
```
{% assign grouptags = ["tag A", "tag A", "Tag B" | ??? %}
```
and I get something like [{"name"=>"tag A",  "count"=>2}{"name"=>"Tag B",  "count"=>1}]

**Comment by ashmaroli** [comment]
Well, there is definitely an unintended change in behavior and therefore inconvenience.

As of now, there is no easy way.
You can use the following code (by saving it into a `*.rb` file inside a directory named `_plugins`) if your GitLab Pages plan allows use of custom plugins:
```ruby
module Jekyll
  module TagFilters
    def group_by_tags(input)
      return input unless groupable?(input)

      groups = input.group_by { |item| item_property(item, "tags") }
      groups.each_with_object([]) do |item, array|
        array << {
          "name"  => item.first.uniq.first,
          "items" => item.last,
          "size"  => item.last.size,
        }
      end
    end
  end
end
```
Then edit your templates to use the custom filter:
```
{% assign alldocs = site.exercises %}
{% assign grouptag =  alldocs | group_by_tags %}
xx {{ grouptag }}
```

If you can't use custom plugins, you may have to switch to using `group_by_exp` filter:

```
{% assign alldocs = site.exercises %}
{% assign grouptag =  alldocs | group_by_exp: 'item', 'item.tags.uniq.first' %}
xx {{ grouptag }}
```

**Comment by moozer** [comment]
@ashmaroli Thanks for your help - this "bug" ended up being support  :-)

Your suggestion about using a plugin made it much easier. It was suddenly possible for me to start searching for ruby solutions, and I ended up at [stackoverflow](https://stackoverflow.com/questions/5128200/how-to-count-identical-string-elements-in-a-ruby-array)

I opted for the more general solution for counting array entries.

```
➜ cat _plugins/grouptag.rb 
module Jekyll
  module TagFilters
    def count_by_item(input)
      counts = Hash.new(0)
      input.each { |name| counts[name] += 1 }
      counts
    end
  end
end


Liquid::Template.register_filter(Jekyll::TagFilters)

```

which results in something like this in jekyll

```
{% assign tag_count = alldocs | map: 'tags' | count_by_item %}

{% for t in tag_count %}
    {{ t[0] }} - {{ t[1] }}
{% endfor %}

```

And apparently when ruby doesn't find a filter, it just ignores it and passes the input data through. Not knowing that made debugging hard.

We have resolved my issue, so closing it.

Thanks again
